File: python/151-160/Min Stack.py
class MinStack:
    stack=[]
    Min=[]
    
    def __init__(self):
        self.stack=[]
        self.Min=[]
    # @param x, an integer
    # @return an integer
    def push(self, x):
        self.stack.append(x)
        if len(self.Min)>0:
            
            t=self.Min[-1]
            if t<=x:
                self.Min.append(t)
            else:
                self.Min.append(x)
        else:
            self.Min.append(x)

    # @return nothing
    def pop(self):
        if len(self.stack)>0:
            del self.stack[-1]
            del self.Min[-1]

    # @return an integer
    def top(self):
        return self.stack[-1]

    # @return an integer
    def getMin(self):
        if len(self.stack)>0:
            return self.Min[-1]

File: Cpp/151-160/MinStack.cpp
/*
    runtime: 40ms
    error : 0
*/

class MinStack {
public:
    
    void push(int x) {
        myStack.push_back(x);
        pos++;
        if(pos>0&&x<Min[pos-1]||pos==0)
            Min.push_back(x);
        else
            Min.push_back(Min[pos-1]);
    }

    void pop() {
        if(pos==-1) return;
        myStack.pop_back();
        Min.pop_back();
        pos--;
    }

    int top() {
        if(pos==-1) return -1;
        return myStack[pos];
    }

    int getMin() {
        if(pos==-1) return -1;
        return Min[pos];
    }
private:
    vector<int> myStack;
    vector<int> Min;
    int pos = -1;
};
